File: 序号/2258. 逃离火灾.kt
class Solution {
  val dirs = arrayOf(intArrayOf(-1, 0), intArrayOf(1, 0), intArrayOf(0, 1), intArrayOf(0, -1))
  fun maximumMinutes(grid: Array<IntArray>): Int {
    // 0草 1\U0001f525 2\U0001f9f1，返回停留最长时间后还能抵达右下角。到不了返回-1，停多久都能到，返回1e9
    val m = grid.size
    val n = grid[0].size
    fun bfs(q: ArrayDeque<IntArray>): IntArray {
      val time = Array(m) { IntArray(n) { -1 } }
      var t = 0
      while (q.isNotEmpty()) {
        val sz = q.size
        for (i in 0 until sz) {
          val (x, y) = q.removeFirst()
          time[x][y] = t
          for (dir in dirs) {
            val newX = x+dir[0]
            val newY = y+dir[1]
            if (newX in 0 until m && newY in 0 until n && grid[newX][newY] == 0 && time[newX][newY] == -1) {
              q.add(intArrayOf(newX, newY))
            }
          }

        }
        t++
      }
      return intArrayOf(time[m - 1][n - 1], time[m - 2][n - 1], time[m - 1][n - 2])
    }
    // 统计火和人到达每个点的最快时间，人更快到房间即可。不会出现火截断人通路的情况，因为火如果截断了，火必定比人更早到房间
    val (a1, b1, c1) = bfs(ArrayDeque<IntArray>().apply { add(intArrayOf(0, 0)) })
    val burns = ArrayDeque<IntArray>()
    for (x in 0 until m) {
      for (y in 0 until n) {
        if (grid[x][y] == 1) {
          burns.add(intArrayOf(x, y))
        }
      }
    }
    val (a2, b2, c2) = bfs(burns)
    if (a1 == -1) return -1 // 人到不了终点
    if (a2 == -1) return 1000000000 // 火到不了终点
    val ans = a2 - a1 // 火比人晚到的时间，就是人可以等待的最大时间
    if (ans < 0) return -1 // 人到终点的时间更长
    if (b1 == -1 || c1 == -1 || 
      (b2 - b1 == ans && c2 - c1 == ans) // 只有 与运算才能保证 人火一路
    ) {
      // 一条路封住了，或者 最后一格重合了
      return ans - 1
    }
    return ans
  }
}

File: 序号/2258. 逃离火灾.py
class Solution:
    def maximumMinutes(self, grid: List[List[int]]) -> int:
        # bfs统计人与火到达每个格子的时间，人比火快多少就是可以等待的最大时间
        # 火不可能中途拦截，反证：假设人到达A点需要3分，火到达A需要4分。
        # 若火在A之前的B点拦截了人，那么在火到达B也一定与人的时间相同，这与假设矛盾
        # 统计到达终点，以及左边 上边两个点的时间。因为题目允许人火同到终点，需要分别计算到达终点两条路的时间
        m,n=len(grid),len(grid[0])
        dirs=(-1,0,1,0,-1)
        def bfs(q:deque):
            times=[[-1]*n for _ in range(m)]
            t=0
            while q:
                sz=len(q)
                for _ in range(sz):
                    x,y=q.popleft()
                    times[x][y]=t
                    for i in range(4):
                        dx,dy=dirs[i],dirs[i+1]
                        nx,ny=x+dx,y+dy
                        if 0<=nx<m and 0<=ny<n and grid[nx][ny]==0 and times[nx][ny]<0:
                            q.append((nx,ny))
                t+=1
            return times[-1][-1],times[-2][-1],times[-1][-2]
        
        fires=deque()
        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    fires.append((i,j))
        pa,pb,pc=bfs(deque([(0,0)]))
        fa,fb,fc=bfs(fires)
        d=fa-pa
        if pa==-1:return -1 # 人到不了
        if fa==-1:return int(10**9) # 火到不了，可以无限等
        if d<0:return -1 # 火比人先到终点，寄
        if pb!=-1 and pb+d<fb:return d # 终点上方的路 人可以走的通，且加上等待的时间都比火先到，可以等d分钟
        if pc!=-1 and pc+d<fc:return d # 终点左边的路 。。。
        return d-1 # 人火一路，人得快一分钟